src/ring_theory/dedekind_domain.lean

```diff
@@ -950,4 +950,586 @@ begin
+  obtain ⟨c, hc⟩ := equiv.surjective (ideal_correspondence hI hJ f)
+    ⟨b, dvd_of_mem_normalized_factors hb⟩,
+  obtain ⟨c, hd⟩ := c,
+  have : (ideal_correspondence hI hJ f ⟨c, hd⟩ : ideal S) = b,
+    rw subtype.coe_eq_iff,
+    use dvd_of_mem_normalized_factors hb,
+    exact hc,
+  rw [dvd_iff_le, ← this, subtype.val_eq_coe, ← ideal_correspondence_mono' hI hJ f
+    (dvd_of_mem_normalized_factors hp) hd] at H,
+  have pmax : p.is_maximal,
+  sorry, --should use a result like is_prime.is_maximal. I've already proven this so I'll add it soon
+  suffices H'' : c ≠ ⊤,
+  { simp only [is_maximal.eq_of_le pmax H'' H, hb, this] },
+  sorry
+end
+
+lemma ideal.dvd_is_prime_pow {p q : ideal T} (hp : p.is_prime) {n : ℕ} :
+  q ∣ p^n ↔ ∃ i ≤ n, q = p^i :=
+sorry
+
+lemma dvd_prime_pow {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M} (hp : prime p) (n : ℕ) :
+  q ∣ p^n ↔ ∃ i ≤ n, associated q (p ^ i) :=
+begin
+  induction n with n ih generalizing q,
+  { simp [← is_unit_iff_dvd_one, associated_one_iff_is_unit] },
+  split,
+  { intro h,
+    rw pow_succ at h,
+    rcases hp.left_dvd_or_dvd_right_of_dvd_mul h with (⟨q, rfl⟩ | hno),
+    { rw [mul_dvd_mul_iff_left hp.ne_zero, ih] at h,
+      rcases h with ⟨i, hi, hq⟩,
+      { refine ⟨i + 1, nat.succ_le_succ hi, (hq.mul_left p).trans _⟩,
+        rw pow_succ } },
+    { obtain ⟨i, hi, hq⟩ := ih.mp hno,
+      exact ⟨i, hi.trans n.le_succ, hq⟩ } },
+  { rintro ⟨i, hi, hq⟩,
+    exact hq.dvd.trans (pow_dvd_pow p hi) },
+end
+
+lemma pow_prime₁' {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M} (n : ℕ) (c : ℕ → M)
+  (h₁ : ∀ i j, i < j → dvd_not_unit (c i) (c j) )
+  (h₂ : ∀ {r : M}, r ∣ q ↔ ∃ i ≤ n,  associated r (c i)) : is_unit (c 0) :=
+sorry
+
+lemma pow_prime₁ {q : ideal T} (n : ℕ) (c : ℕ → ideal T)
+  (h₁ : strict_anti c) (h₂ : ∀ {r : ideal T}, r ∣ q ↔ ∃ i ≤ n, r = c i) :
+  c 0 = ⊤ :=
+begin
+  obtain ⟨i, hi, hr⟩ := h₂.mp (one_dvd _),
+  rw [eq_top_iff, ← ideal.one_eq_top, hr],
+  exact h₁.antitone i.zero_le
+end
+
+lemma not_unit_of_dvd_not_unit {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M}
+  (hp : dvd_not_unit p q): ¬ is_unit q := sorry
+
+lemma not_prime_of_not_unit_dvd_not_unit {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M}
+  (hp : ¬is_unit p)(h : dvd_not_unit p q) : ¬ irreducible q :=
+begin
+  by_contra hcontra,
+  obtain ⟨hp', x, hx, hx'⟩ := h,
+  exact hp (or.resolve_right ((irreducible_iff.1 hcontra).right p x hx') hx),
+end
+
+lemma pow_prime₂' {M : Type*} [comm_cancel_monoid_with_zero M] {p q : M} (n : ℕ) (c : ℕ → M)
+  (h₁ : ∀ i j, i < j → dvd_not_unit (c i) (c j))
+  (h₂ : ∀ {r : M}, r ∣ q ↔ ∃ i ≤ n, associated r (c i)) : prime (c 1) :=
+begin
+  rw prime,
+  split,
+  { sorry },
+  { split,
+    { exact not_unit_of_dvd_not_unit (h₁ 0 1 zero_lt_one) },
+    { sorry } }, --the last part of the proof seems quite painful in the
+                -- `comm_cancel_monoid_with_zero M` setting
+end
```

Comment:
Here's my approach. Then later we can assume `unique_factorization_monoid M` to show that `irreducible (c 1) → prime (c 1)`.
```suggestion
lemma associates.is_atom_iff {M : Type*} [comm_cancel_monoid_with_zero M]
  {p : associates M} : is_atom p ↔ irreducible p :=
⟨λ hp, ⟨by simpa only [associates.is_unit_iff_eq_one] using hp.1,
        λ a b h, (eq_bot_or_eq_of_le_atom hp ⟨_, h⟩).cases_on
          (λ ha, or.inl (a.is_unit_iff_eq_one.mpr ha))
          (λ ha, or.inr sorry)⟩,
 λ hp, ⟨by simpa only [associates.is_unit_iff_eq_one, associates.bot_eq_one] using hp.1,
        λ b ⟨⟨a, hab⟩, hb⟩, (hp.is_unit_or_is_unit hab).cases_on
          (λ hb, show b = ⊥, by rwa [associates.is_unit_iff_eq_one, ← associates.bot_eq_one] at hb)
          (λ ha, absurd (show p ∣ b, from ⟨(ha.unit⁻¹ : units _), by simp [hab]; sorry⟩) hb)⟩⟩

lemma pow_prime₂' {M : Type*} [comm_cancel_monoid_with_zero M] {p q : associates M} (n : ℕ) (hn : 1 ≤ n)
  (c : ℕ → associates M) (h₁ : strict_mono c)
  (h₂ : ∀ {r}, r ∣ q ↔ ∃ i ≤ n, r = (c i)) : irreducible (c 1) :=
begin
  refine associates.is_atom_iff.mp _,
  split,
  { exact ne_bot_of_gt (h₁ zero_lt_one) },

  intros b hb,
  have h : b ∣ q := (hb.le).trans (h₂.2 ⟨1, hn, rfl⟩),
  obtain ⟨i, hi, rfl⟩ := h₂.1 h,
  have H : i < 1 := h₁.lt_iff_lt.mp hb,
  rw [nat.lt_one_iff.mp H, associates.bot_eq_one, ← associates.is_unit_iff_eq_one],
  exact pow_prime₁' n c h₁ @h₂
end
```